### PR TITLE
Combo feature overhaul

### PR DESCRIPTION
**Pull request by Anton3**
General Combo improvements.

## Description
Combos are now defined in a new, backwards-incompatible way. `constexpr` from C++ is utilized to improve performance by an order of magnitude, while reducing RAM requirements to zero. Now all the per-combo data is stored in PROGMEM. There is no repetition in combo definitions, which removes the need for workarounds like germ's `g/...` macros.

Combo processing is moved before TMK's action tapping. This makes combos usable with Action Tap (layer-tap and mod-tap) keys. Also, combos now work on key positions, not on keycodes, which allows for cleaner layer handling. 

All QMK keycodes can now be the result of combos, and combos can now be composed of keys with any keycodes.

Combos now don't fire immediately when a complete combo is detected. Instead the whole `COMBO_TERM` (now defaults to 50) is waited. This allows to define overlapping combos and have only one of the combos trigger. On the other hand, if waiting the full combo term is undesirable, one can use the new "urgent combos" feature.

Combos can now emit not only keycodes, but also strings and function calls.

[Check the docs](https://github.com/qmk/qmk_firmware/blob/c8c353abeb241c7fe72e1401fd5c46bc5a8647ba/docs/feature_combo.md) for the full review.

## Issues

- No extensive testing has been performed yet
- On my keyboard, the new combos weigh about 2KB more than the old ones (LTO on)

## Types of Changes
* [x]  Core
* [x]  Bugfix
* [x]  New feature
* [x]  Enhancement/optimization
* [ ]  Keyboard (addition or update)
* [ ]  Keymap/layout/userspace (addition or update)
* [x]  Documentation

## Issues Fixed or Closed by This PR
* #8003
* #8212

## Checklist
* [x]  My code follows the code style of this project.
* [x]  My change requires a change to the documentation.
* [x]  I have updated the documentation accordingly.
* [x]  I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
* [ ]  I have added tests to cover my changes.
* [ ]  I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Anton3** [comment]
What? How can I cleanup the PR from foreign changes?

**Comment by drashna** [comment]
This needs a rebase.

**Comment by tzarc** [comment]
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

File: tmk_core/avr.mk [hunk 1]
```diff
@@ -28,7 +28,7 @@ CFLAGS += -fno-inline-small-functions
 CFLAGS += -fno-strict-aliasing
 
 CXXFLAGS += $(COMPILEFLAGS)
-CXXFLAGS += -fno-exceptions -std=c++11
+CXXFLAGS += -fno-exceptions -std=c++14
```
Comment (drashna) [inline tmk_core/avr.mk:31]: Is this change necessary? 

And if it is, does it compile on avr-gcc 5.4?

Comment (Anton3) [reply tmk_core/avr.mk:31]: Full-blown `constexpr` starts with C++14, so yes.

It does compile on 5.4, that's what I use.

**Comment by tzarc** [comment]
Unfortunately we're going to have to skip this at this stage.

To put it bluntly, the number of QMK collaborators with the time and appropriate knowledge of C++ for maintaining this (and explaining the implementation to others) doesn't inspire confidence that this will be workable in the long term.

Apologies for dragging this on, and whilst we appreciate the amount of work that went into this, it's just not feasible to pull it in at this stage. Sorry!
